File: solutions/1487-cinema-seat-allocation/solution.py
from collections import defaultdict

class Solution:
    def maxNumberOfFamilies(self, n: int, reservedSeats: list[list[int]]) -> int:
        rows = defaultdict(int)

        # Store reserved seats (only seats 2-9 matter)
        for r, s in reservedSeats:
            if 2 <= s <= 9:
                rows[r] |= 1 << (s - 2)

        # Rows without reservations can fit 2 groups
        ans = (n - len(rows)) * 2

        # Bit masks for the three possible seat blocks
        left = int("00001111", 2)    # seats 2-5
        middle = int("00111100", 2)  # seats 4-7
        right = int("11110000", 2)   # seats 6-9

        for mask in rows.values():
            left_ok = (mask & left) == 0
            middle_ok = (mask & middle) == 0
            right_ok = (mask & right) == 0

            if left_ok and right_ok:
                ans += 2
            elif left_ok or middle_ok or right_ok:
                ans += 1

        return ans

File: solutions/1487-cinema-seat-allocation/solution.java
import java.util.*;

class Solution {
    public int maxNumberOfFamilies(int n, int[][] reservedSeats) {

        // Store reserved seats for each row
        Map<Integer, Set<Integer>> rows = new HashMap<>();

        for (int[] seat : reservedSeats) {
            int row = seat[0];
            int col = seat[1];

            rows.computeIfAbsent(row, k -> new HashSet<>()).add(col);
        }

        // Rows with no reservations can always fit 2 groups
        int result = (n - rows.size()) * 2;

        // Process only rows that have reserved seats
        for (Set<Integer> seats : rows.values()) {

            boolean left = true;   // 2,3,4,5
            boolean middle = true; // 4,5,6,7
            boolean right = true;  // 6,7,8,9

            // Check LEFT block
            for (int i = 2; i <= 5; i++) {
                if (seats.contains(i)) {
                    left = false;
                    break;
                }
            }

            // Check MIDDLE block
            for (int i = 4; i <= 7; i++) {
                if (seats.contains(i)) {
                    middle = false;
                    break;
                }
            }

            // Check RIGHT block
            for (int i = 6; i <= 9; i++) {
                if (seats.contains(i)) {
                    right = false;
                    break;
                }
            }

            if (left && right) {
                result += 2;
            } else if (left || middle || right) {
                result += 1;
            }
        }

        return result;
    }
}
